google/cloud/aiplatform/experimental/vertex_model/base.py

```diff
@@ -172,32 +244,123 @@ def f(*args, **kwargs):
+            )
 
-        # Assuming only local prediction for now
-        output_dir = obj._model._gca_resource.artifact_uri
-        model_uri = pathlib.Path(output_dir) / (
-            "my_" + obj.training_mode + "_model.pth"
-        )
+            # Upload model w/ command
+            import_lines = ""
+            try:
+                module = inspect.getmodule(obj.__class__)
+                import_lines = "\n".join(source_utils.get_import_lines(module.__file__))
+
+            except AttributeError:
+                import_lines = "\n".join(
+                    source_utils.get_import_lines(
+                        source_utils.jupyter_notebook_to_file()
+                    )
+                )
+
+            training_source = source_utils._make_class_source(obj)
+            class_args = inspect.signature(obj.__class__.__init__).bind(
+                obj, *obj._constructor_arguments[0], **obj._constructor_arguments[1]
+            )
+
+            class_creation = f"original_model = {obj.__class__.__name__}({','.join(map(str, class_args.args[1:]))})\n"
+            command_str = (
+                import_lines
+                + COMMAND_STRING_CODE_SETUP
+                + training_source
+                + class_creation
+                + COMMAND_STRING_CODE_APIS
+            )
+
+            obj._model = aiplatform.Model.upload(
+                display_name="serving-test",
+                artifact_uri=vertex_model_model_folder,
+                serving_container_image_uri="gcr.io/google-appengine/python",
+                serving_container_command=COMMAND_STRING_CLI + [command_str],
+            )
+
+        # Cloud training to local prediction: deserialize from cloud URI
+        if not method.__self__.remote:
+            output_dir = obj._model._gca_resource.artifact_uri
+            model_uri = output_dir + "/" + "my_" + "local" + "_model.pth"
 
-        my_model = model._deserialize_remote_model(str(model_uri))
-        return my_model.predict(*args, **kwargs)
+            my_model = model._deserialize_remote_model(model_uri)
+
+            try:
+                my_model.predict(*args, **kwargs)
+            except AttributeError:
+                my_model.predict = functools.partial(obj.__class__.predict, my_model)
+
+            return my_model.predict(*args, **kwargs)
+
+        # Make remote predictions, regardless of training: create custom container
+        if method.__self__.remote:
+            # Convert the predict input to a JSON input for the Endpoint resource
+            data = []
+            bound_args = inspect.signature(method).bind(*args, **kwargs)
+
+            for parameter_name, parameter in bound_args.arguments.items():
+                if parameter_name == "data":
+                    data = parameter.tolist()
+                    break
+
+            # TODO: cleanup model resource after endpoint is created
+            endpoint = obj._model.deploy(machine_type="n1-standard-4")
```

Comment:
Instead of deploying at every predict call we should store the endpoint as an attribute.

```Python
if self._endpoint is None:
  self._endpoint = obj._model.deploy(...)

self._endpoint.predict(instances=data)
```

The attribute should be cleared whenever the user launches a new training as the underlying model has changed.

Reply:
Log that the model will be deployed. Something like: "Model is not deployed for remote prediction. Deploying model to an endpoint."